thehive/app/org/thp/thehive/controllers/v1/PatternCtrl.scala

```diff
@@ -0,0 +1,127 @@
+        val file: FFile = request.body("file")
+
+        for {
+          inputPatterns <- parseJsonFile(file)
+          richPatterns =
+            inputPatterns
+              .sortBy(_.external_id.length) // sort to create sub-patterns after their parent
+              .foldLeft[JsArray](JsArray.empty) { (array, inputPattern) =>
+                val res = db.tryTransaction { implicit graph =>
+                  createFromInput(inputPattern)
+                } match {
+                  case Failure(e) =>
+                    Json.obj("status" -> "Failure", "message" -> e.getMessage)
+                  case Success(t) =>
+                    Json.obj("status" -> "Success", "mitreId" -> t.patternId, "patternName" -> t.name)
+                }
+                array :+ res
+              }
+        } yield Results.Created(richPatterns)
+      }
+
+  def get(patternId: String): Action[AnyContent] =
+    entrypoint("get pattern")
+      .authRoTransaction(db) { implicit request => implicit graph =>
+        patternSrv
+          .get(EntityIdOrName(patternId))
+          .richPattern
+          .getOrFail("Pattern")
+          .map(richPattern => Results.Ok(richPattern.toJson))
+      }
+
+  def getCasePatterns(caseId: String): Action[AnyContent] =
+    entrypoint("get case patterns")
+      .authRoTransaction(db) { implicit request => implicit graph =>
+        for {
+          patternIds <- patternSrv.getCasePatterns(caseId)
+        } yield Results.Ok(patternIds.toJson)
+      }
+
+  def delete(patternId: String): Action[AnyContent] =
+    entrypoint("delete pattern")
+      .authPermittedTransaction(db, Permissions.managePattern) { implicit request => implicit graph =>
+        patternSrv
+          .getOrFail(EntityIdOrName(patternId))
+          .flatMap(patternSrv.remove)
+          .map(_ => Results.NoContent)
+      }
+
+  private def parseJsonFile(file: FFile): Try[Seq[InputPattern]] =
+    for {
+      stream <- Try(new FileInputStream(file.filepath.toString))
+      json = Json.parse(stream)
+    } yield (json \ "objects").get.as[Seq[InputPattern]]
+
+  private def createFromInput(inputPattern: InputPattern)(implicit graph: Graph, authContext: AuthContext): Try[Pattern with Entity] =
+    if (inputPattern.external_id.isEmpty)
+      Failure(BadRequestError(s"A pattern with no MITRE id cannot be imported"))
+    else if (patternSrv.startTraversal.alreadyImported(inputPattern.external_id))
+      Failure(BadRequestError(s"A pattern with MITRE id '${inputPattern.external_id}' already exists in this organisation"))
+    else
+      for {
+        pattern <- patternSrv.createEntity(inputPattern.toPattern)
+        _ = if (inputPattern.x_mitre_is_subtechnique.getOrElse(false)) linkPattern(pattern)
+      } yield pattern
+
+  private def linkPattern(child: Pattern with Entity)(implicit graph: Graph, authContext: AuthContext): Try[Unit] = {
+    val firstDot = child.patternId.indexOf(".")
```

Comment:
Is it possible that patternId doesn't contain a dot ?

Reply:
Normally not but we could do :
```
    val firstDot = child.patternId.indexOf(".")
    for {
      parentId <- Try(child.patternId.substring(0, firstDot))
      parent   <- patternSrv.startTraversal.getByPatternId(parentId).getOrFail("Pattern")
      _        <- patternSrv.setParent(child, parent)
    } yield ()
    ```

Reply:
In that case the controler could generate a `java.lang.StringIndexOutOfBoundsException`, which is unreadable for the end-user. I'd prefer something like `if (firstDot == -1) Failure(BadRequestError("Invalid patternId ${child.patternId} (must contain a dot")) else for { `